File: SeatingPlaceSuggestions/Java/SeatingPlaceSuggestions/src/main/java/org/weaveit/seatssuggestions/AuditoriumSeating.java
package org.weaveit.seatssuggestions;

import java.util.HashMap;
import java.util.Map;

public class AuditoriumSeating {

    private Map<String, Row> rows;

    public AuditoriumSeating(Map<String, Row> rows) {
        this.rows = rows;
    }

    public SeatingOptionSuggested suggestSeatingOptionFor(int partyRequested, PricingCategory pricingCategory) {
        for (Row row : rows.values()) {
            SeatingOptionSuggested seatingOptionSuggested = row.suggestSeatingOption(partyRequested, pricingCategory);

            if (seatingOptionSuggested.matchExpectation()) {
                // Cool, we mark the seat as Allocated (that we turns into a SuggestionMode)
                return seatingOptionSuggested;
            }
        }

        return new SeatingOptionNotAvailable(partyRequested, pricingCategory);
    }

    public AuditoriumSeating allocate(SeatingOptionSuggested seatingOptionSuggested) {
        // Update the seat references in the Auditorium
        Map<String, Row> newVersionOfRows = new HashMap<>(rows);

        for (SeatingPlace updatedSeat : seatingOptionSuggested.seats()) {
            Row formerRow = newVersionOfRows.get(updatedSeat.rowName());
            Row newVersionOfRow = formerRow.allocate(updatedSeat);
            newVersionOfRows.put(updatedSeat.rowName(), newVersionOfRow);
        }

        rows = newVersionOfRows;

        return this;
    }

}

File: SeatingPlaceSuggestions/CSharp/SeatingPlaceSuggestions/AuditoriumSeating.cs
using System.Collections.Generic;
using Value;
using Value.Shared;

namespace SeatsSuggestions;

public class AuditoriumSeating(Dictionary<string, Row> rows) : ValueType<AuditoriumSeating>
{
    public IReadOnlyDictionary<string, Row> Rows => rows;

    public SeatingOptionSuggested SuggestSeatingOptionFor(SuggestionRequest suggestionRequest)
    {
        foreach (var row in rows.Values)
        {
            var seatingOption = row.SuggestSeatingOption(suggestionRequest);

            if (seatingOption.MatchExpectation()) return seatingOption;
        }

        return new SeatingOptionNotAvailable(suggestionRequest);
    }

    public AuditoriumSeating Allocate(SeatingOptionSuggested seatingOptionSuggested)
    {
        // Update the seat references in the Auditorium
        var newVersionOfRows = new Dictionary<string, Row>(rows);

        foreach (var updatedSeat in (IEnumerable<SeatingPlace>)seatingOptionSuggested.Seats)
        {
            var formerRow = newVersionOfRows[updatedSeat.RowName];
            var newVersionOfRow = formerRow.Allocate(updatedSeat);
            newVersionOfRows[updatedSeat.RowName] = newVersionOfRow;
        }

        //rows = newVersionOfRows;

        return  new AuditoriumSeating(newVersionOfRows);
    }

    protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
    {
        return new object[] { new DictionaryByValue<string, Row>(rows) };
    }
}